[REVIEW] Implement cudf::label_bins()

vyasr opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on main/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against main they should be resolved by merging main
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

This PR resolves #7517, implementing a binning feature in `libcudf` to support [pandas.cut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html) in `cudf`.

vyasr commented:
@kkraus14 @jrhemstad tagging you two since I'm not sure who should be reviewing this, feel free to add others who would be helpful.

jrhemstad commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,51 @@
+
+    // Handle empty inputs.
+    if (input.is_empty()) {
+        return cudf::make_numeric_column(data_type(type_id::UINT32), 0);
```
Use `empty_like(input)`  here.

vyasr replied:
I originally used `empty_like` here, but `input` could be of any `dtype` (e.g. floating point numbers) and the return type is always integral, right? Unless I misunderstood your proposed API, I'm just returning a column of indices into `left_edges`.

jrhemstad commented on cpp/include/cudf/binning/bin.cuh:
```diff
@@ -0,0 +1,162 @@
+constexpr unsigned int MYNULL = 0xffffffff;
+
+namespace detail {
+namespace {
```
Don't use anonymous namespaces in headers. https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file

I don't think any of this needs to be in a header. I'd put all of this code inside `bin.cu` inside an anon namespace.

vyasr replied:
You're OK with putting templates into the source file, I assume because in this case we can safely assume that none of these will be used in any other translation unit?

jrhemstad replied:
Right, if the visibility is only limited to that TU with an anon namespace, then it's not possible for it to be used elsewhere.

vyasr replied:
I've done this, just leaving the thread since I'm curious to hear your thoughts re: templates in source files, since I usually avoid that.

jrhemstad replied:
I put templates in source files all the time if I'm only ever going to use that template in that TU.

harrism commented:
Hi @vyasr , thanks for the PR! 

Regarding naming / namespaces / placement of files: I think cudf/binning/bin.hpp is right for the main header. Public API namespace should be `cudf`, and implementation details in `cudf::detail`. Creating a `binning` namespace really only makes sense if there's a lot of stuff in it (e.g. like `cudf::strings`).  

This is all discussed in the [libcudf Developer Guide](https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/docs/DEVELOPER_GUIDE.md)

(Note: Draft PRs aren't assigned reviewers automatically (by design)).

jrhemstad commented on cpp/include/cudf/binning/bin.cuh:
```diff
@@ -0,0 +1,162 @@
+// would be a terribly hacky solution), the current index is not provided to
+// the call operator so I wouldn't know what bit to set (and in any case that
+// would not be parallel-safe without atomics).
+constexpr unsigned int MYNULL = 0xffffffff;
```
I think using a sentinel value for the bin label is a good solution. Then you can use `valid_if` to materialize a bitmask corresponding to the sentinel values.

https://github.com/rapidsai/cudf/blob/c929ba1fe85c152d6e8b4c868cd36f0802dafa51/cpp/include/cudf/detail/valid_if.cuh#L94

jrhemstad commented on cpp/include/cudf/binning/bin.cuh:
```diff
@@ -0,0 +1,162 @@
+
+    thrust::transform(thrust::device,
+            input.begin<T>(), input.end<T>(),
+            static_cast<cudf::mutable_column_view>(*output).begin<size_type>(),
```
```suggestion
            output->mutable_view().begin<size_type>(),
```

jrhemstad commented on cpp/include/cudf/binning/bin.cuh:
```diff
@@ -0,0 +1,162 @@
+        return (m_right_comp(value, m_right_edges[index])) ? index : MYNULL;
+    }
+
+    const T *m_left_edges{};  // Pointer to the beginning of the device data containing left bin edges.
+    const T *m_left_edges_end{};  // Pointer to the end of the device data containing left bin edges.
+    const T *m_right_edges{};  // Pointer to the beginning of the device data containing right bin edges.
```
Use `device_span`.

harrism commented:
I also am not sure about the parameter term "edges". That makes me think of graph or geometric data structures.  Are the bins contiguous? Or do we need to be able to support disjoint bins? (what does Pandas expect?)

vyasr commented:
@harrism thanks for the pointer on the namespaces, sorry I missed that in the developer guide! I'll use`cudf::detail` for internal details and move the public function to `cudf::bin`. 

I intentionally didn't add reviewers since I didn't think this was ready for review and I didn't know who would be the right people to ask questions to, so I just requested review from a subset of you, Kieth, and Jake and figured you would suggest appropriate reviewers. Is that OK for now? Once I have a better idea of who reviews what I can request specific people more easily.

Yes, as @kkraus14 pointed out [here](https://github.com/rapidsai/cudf/issues/7517#issuecomment-792886583) we need to support noncontiguous intervals, which is why we have this slightly awkward API with both left and right edges specified. We could be more explicit and name them `left_bin_edges` and `right_bin_edges`, if you think that would reduce confusion.

jrhemstad commented on cpp/src/binning/bin.cu:
```diff
@@ -16,17 +16,179 @@
+};
+
+/// Bin the input by the edges in left_edges and right_edges.
+template <typename T, typename StrictWeakOrderingLeft, typename StrictWeakOrderingRight, bool InputIsNullable>
```
The naming here implies the comparison operator is a strict, weak ordering operator, but less_equal/greater_equal aren't strict, weak ordering. Does it really need to be a strict, weak ordering op?

jrhemstad replied:
When it doubt, I usually look at the docs for the equivalent STL algorithm. Looking at `[std::lower_bound`](https://en.cppreference.com/w/cpp/algorithm/lower_bound), it looks like the comparator need not be strict. In fact, it need not even satisfy the "Compare" concept. 

So I think you're safe in using those operators, I'd just rename them to avoid confusion (mainly mine).

ttnghia commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,222 @@
+                            column_view const& right_edges,
+                            rmm::mr::device_memory_resource * mr)
+{
+    auto output = cudf::make_numeric_column(data_type(type_to_id<size_type>()), input.size());
```
Please enforce consistency for the namespace prefix `cudf::`. Either you don't add it at all, or add it to everywhere else in your code, such as `cudf::data_type`, `cudf::type_to_id` etc.

vyasr replied:
Thanks for the tip. In general rather than maintain consistency within this file I've tried to follow the most common usage of different identifiers seen in other parts of the code. So far as I can see, there is very minimal consistency in this fashion in other files, preference seems to be given to only explicitly namespace things that are not commonly used (e.g. `device_span` in this file) while using the naked identifiers for more common ones (such as `size_type` and `column`). I'll leave this thread open in case anyone else feels strongly about this, but for now I've updated my code to at least be internally consistent for a given type.

ttnghia commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,222 @@
+            // views provided on the edges will always return const types. The
+            // template arguments to `datq` need not specify const since that
+            // const is added as part of the signature.
+            bin_finder<const T, StrictWeakOrderingLeft, StrictWeakOrderingRight>(
```
Anything inside the `detail` namespace should be prefixed by `detail::`.

ttnghia commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,222 @@
+                )
+            );
+
+    auto mask_and_count = cudf::detail::valid_if(
```
Same here, please enforce consistency for the prefix `cudf::`.

ttnghia commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,222 @@
+}
+
+template <typename T>
+constexpr inline auto is_supported_bin_type()
```
As from the cudf developer guide, constness keyword should come after type auto. So, you should write `auto constexpr` instead.

vyasr replied:
Can you show me what you're referring to?

jrhemstad replied:
@ttnghia we don't specify anything in the developer guide about the order of `auto/constexpr` keywords. In fact, I definitely prefer `constexpr` first as it's usually the most important thing about the symbol.

jrhemstad replied:
That said, `constexpr` and `inline` are redundant.

ttnghia commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,222 @@
+  // TODO: Determine what other types (such as fixed point numbers) should be
+  // supported, and whether any of them (like strings) require special
+  // handling.
+  return (cudf::is_numeric<T>() && not std::is_same<T, bool>::value); // || cudf::is_fixed_point<T>();
```
Please enforce consistency for boolean operators. Don't mix `and`, `or`, `not` with `&&`.

ttnghia commented on cpp/src/binning/bin.cu:
Doxygen comments should be wrapped by /** and */

vyasr replied:
Doxygen comments support a wide range of styles, for example see [here](https://www.star.bnl.gov/public/comp/sofi/doxygen/docblocks.html).

ttnghia replied:
Yeah, but it's not rapid doxygen support. I only know that recently too. Please read this: https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/docs/DEVELOPER_GUIDE.md

ttnghia commented on cpp/tests/binning/bin_tests.cu:
```diff
@@ -0,0 +1,139 @@
+
+namespace {
+
+// =============================================================================
```
Thanks for this, but I don't think such decoration comments are allowed :)

ttnghia commented on cpp/tests/binning/bin_tests.cu:
```diff
@@ -0,0 +1,139 @@
+/// A simple struct to be used as a predicate for comparing a sequence to a given value encoded by this struct in algorithms.
+struct equal_value
+{
+    equal_value(unsigned int value)
```
Please use the initialization style for constructor: `equal_value(... value) : m_value(value) {}`

jrhemstad commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,216 @@
+                    // views provided on the edges will always return const types. The
+                    // template arguments to `datq` need not specify const since that
+                    // const is added as part of the signature.
+                    bin_finder<const T, LeftComparator, RightComparator>(
```
Instead of `const T` here, couldn't it just be `T` and then internally `device_span<T const>`?

vyasr replied:
Yes, it could. I just chose the option that required fewer `const`s :) I'll switch it.

jrhemstad commented on cpp/src/binning/bin.cu:
Be careful with compile time here. This is a lot of kernel instantiations.

vyasr replied:
That's true, I wasn't sure how we should prioritize this. The alternative is dispatching more of these at run time, what would you suggest?

jrhemstad replied:
I suspect putting the nullable/null check inside the kernel wouldn't have much impact on performance.

vyasr replied:
I agree, that's fine with me then.

Regarding the check, in general in cudf do we tend to make decisions based on `input.nullable()` or `input.null_count() == 0`? This question is also relevant [here](https://github.com/rapidsai/cudf/pull/7554/files#diff-637d33f0cb45bee09388ad24ce306f6fb277d8ea8acae5bdb8e29d68f4833cf4R198-R202).

vyasr commented:
@jrhemstad next time you look at this, have a look at the header and compare it to what you posted on the original issue. I updated it somewhat to address a few extra edge cases and describe the current behavior. In case you disagree with any of the ways I've addressed those let me know.

jrhemstad commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,215 @@
+    // _greater than_ the value. This is safe because bound == m_left edges
+    // would already have triggered a NULL_VALUE return above, so there's
+    // no risk of negative values or wraparound for -1 here.
+    auto index = bound - m_left_span.begin() - 1;
```
When doing iterator manipulation like this, it better reveals your intent to use these functions. Otherwise, when doing a quick read it's not obvious if this is normal arithmetic or pointer manipulation. 
```suggestion
    auto index = thrust::distance(thrust::prev(bound), m_left_span.begin());
```

jrhemstad commented on cpp/src/binning/bin.cu:
The public function should call a `detail` function that has a stream parameter, then everything internal should use that stream.

ttnghia commented on cpp/tests/binning/bin_tests.cu:
```diff
@@ -221,8 +221,12 @@ TEST(BinColumnTest, TestLeftEdgesWithNullsAfter)
 
-  std::unique_ptr<cudf::column> result =
-    cudf::bin(input, left_edges, cudf::inclusive::NO, right_edges, cudf::inclusive::NO, cudf::null_order::AFTER);
+  std::unique_ptr<cudf::column> result = cudf::bin(input,
```
You can use `auto const result` to shorten your code.

ttnghia commented on cpp/tests/binning/bin_tests.cu:
```diff
@@ -285,16 +293,19 @@ TEST(BinColumnTest, TestBothEdgesWithNullsAfterDifferentAmount)
   fwc_wrapper<float> right_edges{{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, {1, 1, 1, 1, 1, 1, 1, 1, 1, 0}};
```
Can you use typed test instead?

vyasr replied:
Most of the tests you saw were a limited set of tests I used during development. I just rewrote the testing suite in a final pass, and most of the tests are now typed.

ttnghia commented on cpp/tests/binning/bin_tests.cu:
```diff
@@ -404,16 +415,15 @@ TYPED_TEST_CASE(StringBinTestFixture, StringTypes);
 
 TYPED_TEST(StringBinTestFixture, TestStringData)
```
This function is not typed test. A typed test is a test for various types, using `TypeParam` as the input type.

vyasr commented:
@jrhemstad should we consider reordering the arguments so that we can give defaults to the two inclusion flags? While we may need this libcudf API to support the Python cudf API, I have to imagine that most users of this C++ API will want contiguous half-open (or perhaps half-closed) bins. On the other hand, this may not be that important if we don't anticipate any significant consumers other than cudf.

davidwendt commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,265 @@
+  // These device column views are necessary for creating iterators that work
+  // for columns of compound types. The column_view iterators fail for compound
+  // types because they return raw pointers to the start of the data.
+  auto input_device_view       = column_device_view::create(input);
+  auto left_edges_device_view  = column_device_view::create(left_edges);
+  auto right_edges_device_view = column_device_view::create(right_edges);
```
```suggestion
  auto input_device_view       = column_device_view::create(input, stream);
  auto left_edges_device_view  = column_device_view::create(left_edges, stream);
  auto right_edges_device_view = column_device_view::create(right_edges, stream);
```
Need to pass `stream` parameter to these.

davidwendt commented on cpp/include/cudf/detail/bin.hpp:
```diff
@@ -0,0 +1,62 @@
+  column_view const& right_edges,
+  inclusive right_inclusive,
+  null_order edge_null_precedence     = null_order::BEFORE,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource(),
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default);
```
```suggestion
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```
These 2 parameters need to be reversed.

vyasr replied:
Good catch, thanks. I guess the definition has the correct ordering so this was just being silently declared without being defined.

vyasr replied:
Actually I'm not sure why you suggested this. Is this just libcudf convention? The ordering was consistent between the header and the definition in bin.cu.

vyasr replied:
I've changed this back for now. Why would I want `stream` before `mr`? The non-detail API won't have a `stream` parameter, so isn't the natural API for the `detail` API to have `stream` at the end so that it can be transparently added (by default) to the call from `cudf::bin`?

davidwendt replied:
Short version:   just look at every other detail API in libcudf.
The `mr` is usually a mixture of default and non-default while the stream is only default from the external API and should always be available after that.

davidwendt commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,265 @@
+      input_device_view->pair_begin<T, true>(),
+      input_device_view->pair_end<T, true>(),
+      output_mutable_view.begin<size_type>(),
+      bin_finder<T, decltype(left_edges_device_view->begin<T>()), LeftComparator, RightComparator>(
```
I recommend declaring a type for this instead
```
using RandomAccessIterator = decltype(left_edges_device_view->begin<T>();
```

vyasr replied:
Will do.

davidwendt commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,265 @@
+  RandomAccessIterator m_left_end{};    // The end of the range containing the left bin edges.
+  RandomAccessIterator
+    m_right_begin{};  // The beginning of the range containing the right bin edges.
+  size_type
+    m_edge_index_shift;  // The number of elements m_left_begin has been shifted to skip nulls.
```
```suggestion
  size_type const
    m_edge_index_shift;  // The number of elements m_left_begin has been shifted to skip nulls.
```
All of these probably can be declared `const`.

davidwendt commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,265 @@
+  } else {
+    thrust::transform(
+      rmm::exec_policy(stream),
+      input_device_view->pair_begin<T, false>(),
+      input_device_view->pair_end<T, false>(),
```
```suggestion
      input_device_view->begin<T>(),
      input_device_view->end<T>(),
```
Technically if there are no nulls, you can just use the non-pair iterators.
This is just a suggestion. The symmetry is nice. 
The performance is probably about the same for this.

vyasr replied:
Yes, I considered this. In that case I could overload `bin_finder<...>::operator()` to accept either a pair or a single value, and have the pair version call the value version if it's not null. I'll leave this open for other reviewers to comment as well in case they think it would be worthwhile, perhaps @jrhemstad has an opinion.

vyasr replied:
I implemented this and it works fine. The downside is that in the case where there are nulls it introduces an extra function call (from `operator(thrust::pair<T, bool>)` to `operator(T)`), so unless nvcc can reliably optimize out that out by inlining the function this change would slightly slow the null path while speeding up the non-null path. I can benchmark if there's interest. I don't know if it's significant enough to worry about.

davidwendt commented on cpp/include/cudf/binning/bin.hpp:
```diff
@@ -0,0 +1,89 @@
+  column_view const& left_edges,
+  inclusive left_inclusive,
+  column_view const& right_edges,
+  inclusive right_inclusive,
```
I'm not sure who to ask on this but I think you could just accept a single `inclusive` parameter and instruct the caller that it applies to both edges. Seems odd they would have their nulls sorted differently.

vyasr replied:
I think you may have misunderstood the purpose of the `*_inclusive` parameters. These indicate whether a pair of edges represent an open, closed, or half-open interval (e.g. `(0, 1)`, `(0, 1]`, `[0, 1)`, or `[0, 1]`). They don't have anything to do with nulls. We do need separate parameters for the left and right edges as Keith pointed out [here](https://github.com/rapidsai/cudf/issues/7517#issuecomment-792886583).

The `null_precedence` argument below indicates the null sorting, and as you propose it assumes that both sets of edges have the same null ordering.

davidwendt replied:
Yes, I did misunderstand. Sorry about that.

davidwendt commented on cpp/include/cudf/binning/bin.hpp:
```diff
@@ -0,0 +1,89 @@
+
+// TODO: Figure out the best way to document this enum. It needs to be in the
+// detail header to support that API, but that won't be publicly documented and
+// a using declaration doesn't show the values of the enum.
```
You could define it here and include the non-detail header in the detail header, right?

vyasr replied:
Yes, I definitely could. At the time it seemed strange to be including a non-detail header in a file whose contents are exclusively in the detail namespace, but I suppose we do that all the time.

jrhemstad commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,273 @@
+  auto left_end        = left_edges_device_view->end<T>();
+  auto right_begin     = right_edges_device_view->begin<T>();
+
+  if (edge_null_precedence == null_order::BEFORE) {
+    left_begin  = thrust::next(left_begin, null_shift);
+    right_begin = thrust::next(right_begin, null_shift);
+  } else {
+    left_end = thrust::prev(left_end, null_shift);
+  }
```
I'm confused. Why are nulls allowed in the bin edges?

jrhemstad commented on cpp/include/cudf/binning/bin.hpp:
```diff
@@ -0,0 +1,84 @@
+ *   - NaN elements in `input` belong to no bin and their corresponding label is NULL.
+ *   - Bins must be provided in monotonically increasing order, otherwise behavior is undefined.
+ *   - If two or more bins overlap, behavior is undefined.
+ *   - If nulls are interspersed throughout the edges rather than being at the beginning or end
+ *     (as set by @p edge_null_precedence), behavior is undefined.
```
This strikes me as odd. What's the usecase to allow this? If the user knows all of the nulls are contiguous they could do a slice to just give us the non-null elements.

vyasr replied:
The original API spec we were working with didn't specify anything regarding the behavior of the function with respect to nulls in the edges. It was pointed out to me that very few places in cudf _can't_ handle nulls, but also that for sorted inputs the nulls will be contiguously placed at either the beginning or the end of the column, so this seemed like an easy use-case to support. I'd also be fine just throwing a logic error if you think that's a more sensible choice here.

vyasr replied:
Fair enough. That simplifies this code a bit. Should I bother to throw a logic error if the edges contain nulls, or does this fall under "libcudf doesn't make decision" and we just document null edges as leading to UB?

jrhemstad replied:
We always check for nulls in a situation where nulls are not allowed. I can see how one could make the argument that this is like making a decision, but in this case I think it's a good decision.

vyasr replied:
That sounds like a reasonable choice to me.

davidwendt commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,273 @@
+constexpr auto is_supported_bin_type()
+{
+  return ((is_numeric<T>() && !std::is_same<T, bool>::value)) ||
+         std::is_same<T, string_view>::value;
```
I wonder if this could be simply replaced with `cudf::is_relationally_comparable()` and `cudf::is_relationally_equal()` rather listing specific types.

davidwendt commented on cpp/tests/binning/bin_tests.cpp:
```diff
@@ -0,0 +1,429 @@
+
+namespace {
+
+using namespace cudf::test;
```
Is this being used?

vyasr replied:
Not much anymore, but yes. Before I cleaned up this file all the individual tests had their own wrapper declarations etc that required access to cudf::test. Now it's just that the other typedefs just below this become a bit more cumbersome without it.

davidwendt replied:
I'm just not seeing where that is the case in this file. All the types below seem fully-qualified correctly. I would recommend removing this.

vyasr replied:
`Concat` and the various type lists (`FloatingPointTypes` and `FixedPointTypes` in particular are also used in specific tests below) are not consistently qualified. But I'm in favor of not using namespaces where it's not necessary, so I can fully qualify everything and remove this.

davidwendt commented on cpp/tests/binning/bin_tests.cpp:
```diff
@@ -0,0 +1,429 @@
+}
+
+}  // anonymous namespace
+
```
The only other gtests I would recommend is to add some sliced column views. The current code should handle this fine because of the begin/end iterators automatically account for sliced views. It would be good to have 1 or 2 in case the code is changed in the future.

davidwendt commented on cpp/src/binning/bin.cu:
```diff
@@ -0,0 +1,244 @@
+
+  // Handle empty inputs.
+  if (input.is_empty()) {
+    return make_numeric_column(
+      data_type(type_to_id<size_type>()), 0, mask_state::UNALLOCATED, rmm::cuda_stream_default, mr);
```
```suggestion
    return make_empty_column( data_type(type_to_id<size_type>()) );
```

jrhemstad commented:
> @jrhemstad should we consider reordering the arguments so that we can give defaults to the two inclusion flags? While we may need this libcudf API to support the Python cudf API, I have to imagine that most users of this C++ API will want contiguous half-open (or perhaps half-closed) bins. On the other hand, this may not be that important if we don't anticipate any significant consumers other than cudf.

I think it's better to be explicit here. Any choice we make would feel arbitrary. Better to make the caller be explicit.

jrhemstad reviewed (commented):
I unsure about adding the `cudf/binning` directory. Do we expect other "binning" functionality? Maybe histograms? 

That said, I don't have a great suggestion of where else to put it. Maybe `cudf/transform`
?

jrhemstad commented on cpp/include/cudf/binning/bin.hpp:
```diff
@@ -0,0 +1,82 @@
+ * @brief Labels elements based on membership in the specified bins.
+ *
+ * A bin `i` is defined by `left_edges[i], right_edges[i]`. Whether the edges are inclusive or
+ * not is determined by `left_inclusive` and `right_inclusive`, respectively. If either
+ * `left_edges[i]` or `right_edges[i]` is NULL, no value in `input` can be placed in bin `i`.
```
This makes it sound like NULLs are allowed in the edges.

vyasr replied:
Ah that's a holdover from when I was supporting nulls. I'll fix that, but I'll hold off until I get a few more reviews so that I can make all the changes together

vyasr commented:
> I unsure about adding the `cudf/binning` directory. Do we expect other "binning" functionality? Maybe histograms?
> 
> That said, I don't have a great suggestion of where else to put it. Maybe `cudf/transform`
> ?

That's probably a question for @harrism or @kkraus14. The only other related pandas function I'm aware of is `pandas.qcut`, which splits into quantiles, i.e. instead of specifying a number of bins or bin locations the user specifies either a number of percentiles or the percentile boundaries. I don't know if implementing that is a priority or even on the roadmap; offhand I'd guess it's something for which we don't have a timeline but which we might do eventually.

Even if we did choose to implement `qcut` eventually, depending on how critical performance is we might implement that purely in the Cython/Python layer of cudf rather than in libcudf. Since it requires determining the quantile values first, and pandas has `DataFrame.quantile`, the simplest implementation of `pandas.qcut` in `cudf` would be implementing something like `cudf.DataFrame.quantile` (AFAIK this doesn't exist yet in cudf) to compute the quantiles and then just calling `cudf::bin` with the predetermined bin edges. Then we wouldn't need any new libcudf code at all. Whether or not that's the direction we'd actually want to take is something @kkraus14 idea about.

harrism commented:
TLDR: I think that the group for this algorithm should be called "labeling". I also think we should consider renaming it to `label_bins.`

First, look at the file https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/include/doxygen_groups.h and see if this new algorithm fits into one of those algorithmic groups.

My first reaction was that building a histogram belongs in the "reordering" group, like sorted_order. But then I read your documentation and realized that this doesn't actually reorder, it just labels the element with their bin ID. This output can't even be directly used to reorder into bins. To do that you would have to use the bin labels as a key in a key-index sort and then use the sorted indices to gather into the new order.  So it's not reordering.

My next thought is that this is similar to the new (upcoming) join APIs which will just return a gather mask. But it's not because it doesn't return a gather mask, just something that could be used (see above) to compute a gather mask.  Even so, this bin membership does sort of look like a type of join.

But in general, when I look at an algorithm I try to generalize what it does so that when we add other algorithms that also do that thing we can group them. It feels like this is a labeling algorithm. And it doesn't actually "bin" the inputs (doesn't reorder them or help you reorder them), it simply labels the elements by their bin IDs.

Another way  to think of it is that it's just a vectorized lower_bound -- for each element, it finds the lower_bound of the element in the sorted array of bin edges.  The twist is that the bins have explicit left and right edges rather than being contiguous.

harrism commented:
Please remember to update the PR title since this is what gets copied into the CHANGELOG.md

jrhemstad commented:
@gpucibot merge
